crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+					for (who, entitlement) in royalties_schedule.entitlements.into_iter() {
+						let amount = entitlement * listing.fixed_price;
+						imbalance = imbalance.offset(T::MultiCurrency::deposit_into_existing(&who, Some(listing.payment_asset), amount)?).map_err(|_| Error::<T>::InternalPayment)?;
+					}
+				}
+
+				// must not fail now that payment has been made
+				Self::transfer_ownership(&collection_id, token_id, &current_owner, &origin);
+				Self::remove_direct_listing(&collection_id, token_id);
+				Self::deposit_event(RawEvent::DirectSaleComplete(collection_id, token_id, origin, listing.payment_asset, listing.fixed_price));
+			} else {
+				return Err(Error::<T>::NotForDirectSale.into());
+			}
+		}
+
+		/// Sell NFT on the open market to the highest bidder
+		/// Caller must be the token owner
+		/// - `reserve_price` winning bid must be over this threshold
+		/// - `payment_asset` fungible asset Id to receive payment with
+		/// - `duration` length of the auction (in blocks), uses default duration if unspecified
+		#[weight = T::WeightInfo::auction()]
+		fn auction(origin, collection_id: CollectionId, token_id: T::TokenId, payment_asset: AssetId, reserve_price: Balance, duration: Option<T::BlockNumber>) {
+			let origin = ensure_signed(origin)?;
+			let current_owner = Self::token_owner(&collection_id, token_id);
+			ensure!(current_owner == origin, Error::<T>::NoPermission);
+
+			ensure!(!<Listings<T>>::contains_key(&collection_id, token_id), Error::<T>::TokenListingProtection);
+
+			let listing_end_block = duration.unwrap_or_else(|| <frame_system::Module<T>>::block_number().saturating_add(T::DefaultListingDuration::get()));
+			ListingEndSchedule::<T>::mutate(listing_end_block, |schedule| schedule.push((collection_id.clone(), token_id)));
+			let listing = Listing::<T>::Auction(
+				AuctionListing::<T> {
+					payment_asset,
+					reserve_price,
+					close: listing_end_block,
+				}
+			);
+			Listings::insert(&collection_id, token_id, listing);
+
+			Self::deposit_event(RawEvent::AuctionOpen(collection_id, token_id, payment_asset, reserve_price));
+		}
+
+		/// Place a bid on an open auction
+		/// - `amount` to bid (in the seller's requested payment asset)
+		#[weight = T::WeightInfo::bid()]
+		#[transactional]
+		fn bid(origin, collection_id: CollectionId, token_id: T::TokenId, amount: Balance) {
+			let origin = ensure_signed(origin)?;
+			ensure!(<Listings<T>>::contains_key(&collection_id, token_id), Error::<T>::NotForAuction);
+
+			if let Some(Listing::Auction(listing)) = Self::listings(&collection_id, token_id) {
+				if let Some(current_bid) = Self::listing_winning_bid(&collection_id, token_id) {
+					ensure!(amount > current_bid.1, Error::<T>::BidTooLow);
+				} else {
+					// first bid
+					ensure!(amount >= listing.reserve_price, Error::<T>::BidTooLow);
```

Comment:
We only need to do this check when settling the auction. It's good piece of information to the seller if they know there has been a bid though below the reserve_price.

Reply:
can you elaborate on this? the unsuccesful bid will still be logged as an event on the network

Reply:
There is a strategy behind a couple of my suggestions on the NFT module: "Set and apply limitations only if they are absolutely necessary". This would allow flexibilty, scalabilty and new use cases that you haven't thought at the time of developing your code. Interestingly, most of the time, this makes the code be simpler too. Obviously it's not about allowing any fragility in your design or code. This policy would make sense when talking about safe interactions.

Reply:
I get the idea but it's just guessing at this point. seller can set reserve price 0 and have no issues

Reply:
0 would be like they have sold it with no control over the outcome. Anyway, the decision is yours.